### PR TITLE
HLT DQM Cleaning

### PR DESCRIPTION
**Pull request by andrea21z**
#### PR description:

This PR is created to remove unnecessary histograms for DQM HLT Muon validation and adding other interesting. Summary:
    - This is an update of the [PR #37117](https://github.com/cms-sw/cmssw/pull/37117).
    - Keeping only the most representative HLT folders.
    - Removing histograms that are not used in the validation and adding the number of primary vertex distribution, interesting for Run 3.

#### PR validation:
    - I check the output of the new branch running runTheMatrix and everything looks as expected.
    - I execute the basic test suggested in the CMSSW PR instructions.

### CODE REVIEW

**Comment by qliphy** [comment]
please test

File: DQMOffline/Trigger/src/HLTMuonMatchAndPlot.cc [hunk 1]
```diff
@@ -273,12 +279,34 @@ void HLTMuonMatchAndPlot::analyze(Handle<MuonCollection>& allMuons,
...
       passTriggerDZ = passTriggerDZ || (trigNames.triggerName(hltIndex).find(nonDZPath) != std::string::npos &&
                                         triggerResults->wasrun(hltIndex) && triggerResults->accept(hltIndex));
     }
   }
+  if (dzPath && targetMuons.size() > 1 && passTriggerDZ) {
+    const Track* track0 = nullptr;
+    const Track* track1 = nullptr;
+    if (targetMuons.at(0).isTrackerMuon())
+      track0 = &*targetMuons.at(0).innerTrack();
+    else if (targetMuons.at(0).isTrackerMuon())
```
[the remark this answers is not in the document]
Comment (andrea21z) [reply DQMOffline/Trigger/src/HLTMuonMatchAndPlot.cc]: yes, done!

File: DQMOffline/Trigger/src/HLTMuonMatchAndPlot.cc [hunk 2]
```diff
@@ -273,12 +279,34 @@ void HLTMuonMatchAndPlot::analyze(Handle<MuonCollection>& allMuons,
...
+    const Track* track0 = nullptr;
+    const Track* track1 = nullptr;
+    if (targetMuons.at(0).isTrackerMuon())
+      track0 = &*targetMuons.at(0).innerTrack();
+    else if (targetMuons.at(0).isStandAloneMuon())
+      track0 = &*targetMuons.at(0).outerTrack();
+    if (targetMuons.at(1).isTrackerMuon())
+      track1 = &*targetMuons.at(1).innerTrack();
+    else if (targetMuons.at(1).isStandAloneMuon())
+      track1 = &*targetMuons.at(1).outerTrack();
```
Comment (missirol) [inline DQMOffline/Trigger/src/HLTMuonMatchAndPlot.cc:299]:
```suggestion
    auto track_ptr = [](reco::Muon const& muo){
      if (muo.isTrackerMuon())
        return &*muo.innerTrack();
      else if (muo.isStandAloneMuon())
        return &*muo.outerTrack();
      return nullptr;
    }
    auto const* track0 = track_ptr(targetMuons[0]);
    auto const* track1 = track_ptr(targetMuons[1]);
```
Arguably more readable, I have not tested it.

File: HLTriggerOffline/Muon/python/HLTmultiTrackValidatorMuonTracks_cff.py [hunk 3]
```diff
@@ -5,8 +5,8 @@
     label = [
         "hltIter0HighPtTkMuPixelTracks",
         "hltIter0HighPtTkMuTrackSelectionHighPurity",
-        "hltIter2HighPtTkMuTrackSelectionHighPurity",
-        "hltIter2HighPtTkMuMerged",
+        #"hltIter2HighPtTkMuTrackSelectionHighPurity",
+        #"hltIter2HighPtTkMuMerged",
```
Comment (missirol) [inline HLTriggerOffline/Muon/python/HLTmultiTrackValidatorMuonTracks_cff.py:9]:
```suggestion
```
Any reason to keep the commented-out lines? If not, remove.

File: DQMOffline/Trigger/src/HLTMuonMatchAndPlot.cc [hunk 4]
```diff
@@ -273,12 +279,34 @@ void HLTMuonMatchAndPlot::analyze(Handle<MuonCollection>& allMuons,
...
+
+    if (track0 && track1) {
+      hists_["efficiency_DZ_Mu_denom"]->Fill(track0->dz(beamSpot->position()) - track1->dz(beamSpot->position()));
+    }
+    if (nMatched > 1) {
+      if (track0 && track1) {
+        hists_["efficiency_DZ_Mu_numer"]->Fill(track0->dz(beamSpot->position()) - track1->dz(beamSpot->position()));
+      }
+    }
```
Comment (missirol) [inline DQMOffline/Trigger/src/HLTMuonMatchAndPlot.cc:308]:
```suggestion
    if (track0 && track1) {
      hists_["efficiency_DZ_Mu_denom"]->Fill(track0->dz(beamSpot->position()) - track1->dz(beamSpot->position()));
      if (nMatched > 1) {
        hists_["efficiency_DZ_Mu_numer"]->Fill(track0->dz(beamSpot->position()) - track1->dz(beamSpot->position()));
      }
    }
```

**Review by missirol — COMMENTED** [review]
Optional comments.

**Comment by emanueleusai** [comment]
please test

**Comment by emanueleusai** [comment]
type hlt

**Comment by emanueleusai** [comment]
+1

**Comment by perrotta** [comment]
Following the notifications of the static analyzer, I noticed that:
- In HLTMuonMatchAndPlot.cc the variables `passTrigger` and `passTriggerSS` are never used: that's probably wrong, otherwise it would be better do not define and compute them at all.
- The method `fillEdges` does not initialize `nBins` if binning.size() < 3. This can then result in calling `iBooker.book1D`/`2D` with uninitialized values in the arguments. This can probably get fixed by throwing at [L213](https://github.com/cms-sw/cmssw/blob/bf414b7ffab9bc543c16170a5c858608cb47ad41/DQMOffline/Trigger/src/HLTMuonMatchAndPlot.cc#L313) instead of just issuing a warning. Or alternatively the method `fillEdges`  could return a bool that can be used to decide whether the histos have to be booked or not, This is a minor point with respect to the previous one, but since you are cleaning you can also take care of it.

**Comment by qliphy** [comment]
@andrea21z Would you please have a look at above comments?

**Comment by andrea21z** [comment]
> @andrea21z Would you please have a look at above comments?

Sorry, I was in a conference last week. I will try to implement your comments.

**Comment by emanueleusai** [comment]
please test

File: DQMOffline/Trigger/interface/HLTMuonMatchAndPlot.h [hunk 5]
```diff
@@ -73,7 +73,7 @@ class HLTMuonMatchAndPlot {
   void endRun(const edm::Run &, const edm::EventSetup &);
 
   // Helper Methods
-  void fillEdges(size_t &nBins, float *&edges, const std::vector<double> &binning);
+  void fillEdges(size_t &nBins, float *&edges, const std::vector<double> &binning, bool &bookhist);
```
Comment (perrotta) [inline DQMOffline/Trigger/interface/HLTMuonMatchAndPlot.h:76]:
```suggestion
  bool fillEdges(size_t &nBins, float *&edges, const std::vector<double> &binning);
```

File: DQMOffline/Trigger/src/HLTMuonMatchAndPlot.cc [hunk 6]
```diff
@@ -454,20 +457,22 @@ void HLTMuonMatchAndPlot::book2D(DQMStore::IBooker& iBooker,
...
-  hists_[name] = iBooker.book2D(name.c_str(), title.c_str(), nBinsX, edgesX, nBinsY, edgesY);
-  if (hists_[name])
-    if (hists_[name]->getTH2F()->GetSumw2N())
-      hists_[name]->enableSumw2();
-
-  if (edgesX)
-    delete[] edgesX;
-  if (edgesY)
-    delete[] edgesY;
+  fillEdges(nBinsY, edgesY, binParams_[binningTypeY], bookhist);
```
[the remark this answers is not in the document]
Comment (andrea21z) [reply DQMOffline/Trigger/src/HLTMuonMatchAndPlot.cc]: I have implemented all your comments, thank you

**Comment by perrotta** [comment]
please test

**Comment by emanueleusai** [comment]
+1

**Comment by perrotta** [comment]
+1

**Pull request merged**